### PR TITLE
Changed NodeNG.tolineno to use end_lineno when it is available

### LINKED ISSUE #1350: Decorator.toline is off by 1
### Steps to reproduce

I came across this inconsistency while debugging why pylint reports `missing-docstring` on the wrong line for the `g2` function in the example. As it turns out, the `toline` of the decorator seems to point to `b=3,` instead of `)`.

```python
import ast
import astroid

source = """\
@f(a=2,
   b=3,
)
def g2():
    pass
"""

[f] = ast.parse(source).body
[deco] = f.decorator_list
print("ast", deco.lineno, deco.end_lineno)

[f] = astroid.parse(source).body
[deco] = f.decorators.nodes
print("astroid", deco.fromlineno, deco.tolineno)
```

### Current behavior

```
ast 1 3
astroid 1 2
```

### Expected behavior

```
ast 1 3
astroid 1 3
```

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output

2.9.3

### PR DESCRIPTION
**Pull request by superbobry**
## Steps

TODO: Update ChangeLog.

## Description

astroid currently relies on heuristics for computing node ranges. This was necessary prior to 
Python 3.8, because the `ast` module did not report `end_lineno` (nor `end_col_offset`) for 
the parsed nodes. 

This PR changes `NodeNG.toline` to prefer `end_lineno` coming from the `ast` module,
if available, to the one computed heuristically.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Related Issue

Closes #1350.

### CODE REVIEW

File: tests/unittest_builder.py [hunk 1]
```diff
@@ -70,12 +70,12 @@ def test_callfunc_lineno(self) -> None:
         self.assertEqual(name.tolineno, 4)
         strarg = callfunc.args[0]
         self.assertIsInstance(strarg, nodes.Const)
-        if hasattr(sys, "pypy_version_info"):
-            lineno = 4
+        if hasattr(sys, "pypy_version_info") or not PY38_PLUS:
```
Comment (DanielNoord) [inline tests/unittest_builder.py:73]: I believe `PY38_PLUS` also works for `PyPy`. Could you check this? The old `hasattr` might be unnecessary.

File: tests/unittest_scoped_nodes.py [hunk 2]
```diff
@@ -1078,15 +1078,16 @@ def g1(x):
...
             def g2():
                 pass
         """
         astroid = builder.parse(data)
         self.assertEqual(astroid["g1"].fromlineno, 4)
         self.assertEqual(astroid["g1"].tolineno, 5)
-        self.assertEqual(astroid["g2"].fromlineno, 9)
-        self.assertEqual(astroid["g2"].tolineno, 10)
+        self.assertEqual(astroid["g2"].fromlineno, 10)
+        self.assertEqual(astroid["g2"].tolineno, 11)
```
Comment (DanielNoord) [inline tests/unittest_scoped_nodes.py:1090]: Is this change necessary?

**Review by DanielNoord — COMMENTED** [review]
I did a little investigation and `TryFinally` seems to be handled by a special `_elsed_block_range` function. There might be some clues there.

File: tests/unittest_scoped_nodes.py (hunk 2 above)
Comment (superbobry) [reply tests/unittest_scoped_nodes.py:1090]: Yes, this is a reproducer for #1350. I guess I'd have to branch here as well, because on <3.8 the range reported by astroid is (9, 10).

Comment (DanielNoord) [reply tests/unittest_scoped_nodes.py:1090]: Ah sorry, my bad!

**Comment by DanielNoord** [comment]
@superbobry Sorry for pushing to your branch without prior notice. I was playing around with some of the problems you identified and it felt a waste of effort to not push what I had found.

This is actually an interesting PR as the change to use `end_lineno` for `toline` seems relatively straightforward compared to `lineno` for `fromlineno`. @cdce8p mentioned in #1395 that that change broke 50+ tests in `pylint`.

For the change proposed in this PR only three tests in `pylint` need updating. In fact, two of those can actually be considered bug fixes, while one is due to the way I opted to handle the `TryExcept` issue. The fact that we solve two bugs with this is also a reason for me to push this PR to `2.10` and try and get it in.
I have made https://github.com/PyCQA/pylint/pull/5802 to show the necessary changes.

The `TryExcept` issue is something I would like @cdce8p's opinion on as he probably encountered this as well when we added `end_lineno` and might have good reasons to include the `finally` part of a try except.
We could also move this change out of this PR as it is not completely related.

Let me know what you guys think! _I'm undrafting as this now passes the CI and is ready for review_

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for this @superbobry, and @DanielNoord  LGTM.

**Comment by superbobry** [comment]
Thanks for taking the time to investigate and fix this @DanielNoord!

File: astroid/rebuilder.py [hunk 3]
```diff
@@ -2125,11 +2125,20 @@ def visit_starred(self, node: "ast.Starred", parent: NodeNG) -> nodes.Starred:
...
         if sys.version_info >= (3, 8):
+            # TryExcept excludes the 'finally' but that will be included in the
+            # end_lineno from 'node'. Therefore, we iterate over all non 'finally'
+            # children to find the correct end_lineno and column.
+            last_child_end_lineno = node.lineno
+            last_child_end_col_offset = node.end_col_offset
+            for child in node.body + node.handlers + node.orelse:
+                if child.end_lineno > last_child_end_lineno:
+                    last_child_end_lineno = child.end_lineno
+                    last_child_end_col_offset = child.end_col_offset
```
Comment (cdce8p) [inline astroid/rebuilder.py:2136]: Can't we just always use the last child of `node.body + node.handlers + node.orelse`. The `try` statement has a pretty strict structure, i.e. the if there is an `else` block, it's always **after** the `except` ones.

**Review by cdce8p — COMMENTED** [review]
My first thought with `TryExcept` was to exclude it all together. At some point we should replace it anyway we just a `Try` node. #1389

However, technically the current values for `end_lineno` and `end_col_offset` are wrong so it's probably ok to fix them.

File: astroid/rebuilder.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/rebuilder.py]: Not sure what you meant. See the latest commit, is that what you were looking for?

Comment (cdce8p) [reply astroid/rebuilder.py]: Not exactly. I had something like this in mind (if it works)
```py
            end_lineno = node.end_lineno
            end_col_offset = node.end_col_offset
            all_children: List["ast.AST"] = [*node.body, *node.handlers, *node.orelse]
            for child in reversed(all_children):
                end_lineno = child.end_lineno
                end_col_offset = child.end_col_offset
                break
```

**Review by Pierre-Sassoulas — APPROVED** [review]

File: ChangeLog [hunk 4]
```diff
@@ -30,6 +30,11 @@ Release date: TBA
 
   Closes #1330
 
+* Use the ``end_lineno`` attribute for the ``NodeNG.toline`` property
```
Comment (cdce8p) [inline ChangeLog:33]:
```suggestion
* Use the ``end_lineno`` attribute for the ``NodeNG.tolineno`` property
```

**Review by cdce8p — APPROVED** [review]

**Comment by cdce8p** [comment]
@Pierre-Sassoulas I think this is the last PR for `2.10`. I've looked through the milestone list and moved everything that still needs some work (including the `NamedTuple` change @DanielNoord mentioned). Once those are ready, we could also consider releasing `2.11`.

Let me know if I missed anything, but I think you could do the release after this one is merged.

**Comment by DanielNoord** [comment]
> @Pierre-Sassoulas I think this is the last PR for `2.10`. I've looked through the milestone list and moved everything that still needs some work (including the `NamedTuple` change @DanielNoord mentioned). Once those are ready, we could also consider releasing `2.11`.
> 
> Let me know if I missed anything, but I think you could do the release after this one is merged.

Agreed! 👍

**Comment by DanielNoord** [comment]
And thanks @cdce8p for your efforts over this weekend 😄

**Comment by cdce8p** [comment]
Thanks @superbobry and @DanielNoord for the PR 🚀
